### PR TITLE
Books can hide recipes or display alternate recipe names

### LINKED ISSUE #36790: book_learn optional third entry does nothing?
# Describe the bug

Locally, I added a cookbook cookbook_liverforkids. I then added the following line to the foie_gras result of offal_dishes.json:

`"book_learn": [ [ "mag_cooking", 4 ], [ "mag_glam", 5 ], [ "cookbook_liverforkids", 5, "Buttery Duck Bites" ] ]`

The recipe can be learned from the book in game, but the display shows no mention of "Buttery Duck Bites"

# Steps To Reproduce

1. Add a third entry to book_learn identifier.
2. Wish yourself the book using the debug.
3. Read/view the book. There is no mention of the string added in step one.

# Expected behavior

From the doc in JSON_INFO it reads: "The optional third entry defines a name for the recipe as it should appear in the books description"

I took this to mean that the book description would display the recipe as specified in the third entry string. It displays it as the result instead.
cdda-liver

# Screenshots

# Versions and configuration

n/a

# Additional context

at JSON_INFO.md#recipes it reads:

```
"book_learn": [              // (optional) Array of books that this recipe can be learned from. Each entry contains the id of the book and the skill level at which it can be learned.
    [ "textbook_anarch", 7, "something" ], // The optional third entry defines a name for the recipe as it should appear in the books description (default is the name of resulting item of the recipe)
    [ "textbook_gaswarfare", 8, "" ] // If the name is empty, the recipe is hidden, it will not be shown in the description of the book.
],
"difficulty": 3,             // Difficulty of success check
```

```
[ "textbook_anarch", 7, "something" ], // The optional third entry defines a name for the recipe as it should appear in the books description (default is the name of resulting item of the recipe)
```

But the optional third entry doesn't appear to do anything...

In recipe.cpp the only of mention of book_learn is:

```c++
    if( jo.has_member( "book_learn" ) ) {
        booksets.clear();
        for( JsonArray arr : jo.get_array( "book_learn" ) ) {
            booksets.emplace( arr.get_string( 0 ), arr.size() > 1 ? arr.get_int( 1 ) : -1 );
        }
    }
```

And in fact it appears that that is the only mention in all of src. 

So is the third optional entry deprecated? WIP?

### PR DESCRIPTION
**Pull request by mqrause**
#### Summary

SUMMARY: Bugfixes "Optional third parameter in "book_learn" JSON entry can be used (again)"

#### Purpose of change

Fixes #36790

#### Describe the solution

The JSON for "book_learn" got an update to be a map instead of a nested array. The documentation got updated accordingly. The old JSON still works as before, so no alternate names or hidden recipes.

Created a new struct to use in booksets map that holds all the necessary data to get passed into the book description.

#### Testing

Spawned a book with new JSON format.

### CODE REVIEW

File: src/recipe.h [hunk 1]
```diff
@@ -50,6 +51,26 @@ struct recipe_proficiency {
     void deserialize( JsonIn &jsin );
 };
 
+struct book_recipe_data {
+    int skill_req = -1;
+    bool use_alt_name = false;
+    std::string alt_name = "";
+    bool hidden = false;
+
+    book_recipe_data( JsonArray arr ) {
```
Comment (Qrox) [inline src/recipe.h:60]: Please move implementation to the .cpp file. Also, the convention is to read using a `deserialize` function.

File: src/recipe.h [hunk 2]
```diff
@@ -50,6 +51,26 @@ struct recipe_proficiency {
...
+        int arr_size = arr.size();
+        if( arr_size > 1 ) {
```
Comment (Qrox) [inline src/recipe.h:62]: Instead of using a confusing array, consider using an object and read the values using keys.

File: src/recipe.h [hunk 3]
```diff
@@ -50,6 +51,26 @@ struct recipe_proficiency {
...
+            skill_req = arr.get_int( 1 );
+        }
+        if( arr_size > 2 ) {
+            std::string name = arr.get_string( 2 );
```
Comment (Qrox) [inline src/recipe.h:66]: You should use `class translation` here for translatable strings.

File: src/recipe.h [hunk 4]
```diff
@@ -50,6 +51,26 @@ struct recipe_proficiency {
...
+            alt_name = name;
+            use_alt_name = true;
```
Comment (Qrox) [inline src/recipe.h:68]: Use `cata::optional` for optional values.

File: src/recipe.h [hunk 5]
```diff
@@ -50,6 +51,26 @@ struct recipe_proficiency {
...
+            hidden = name.empty();
```
Comment (Qrox) [inline src/recipe.h:69]: Please use a specialized json value for hiding instead of checking the emptiness of the name.

**Comment by Jerimee** [comment]
Thank you mqrause and Qrox! So glad to see this being implemented.

File: src/recipe.h (hunk 5 above)
Comment (mqrause) [reply src/recipe.h:69]: Oof, while I do agree, that would likely mean switching book_learn from arrays to objects and thus probably adjusting all the recipes. I was hoping to get away with just reimplementing what's in the documentation.

Comment (Qrox) [reply src/recipe.h:69]: You can implement a new format while maintaining back-compatibility for the old format.

Comment (mqrause) [reply src/recipe.h:69]: After looking into it a bit more, this seems to be the only occurence of an array with individual optional parameters. It probably really should be replaced by an object, with backwards compatibility where you don't want to use alternate names or hide it.

File: src/recipe.h (hunk 3 above)
Comment (mqrause) [reply src/recipe.h]: I hope I used that correctly.

File: src/recipe.h (hunk 5 above)
Comment (mqrause) [reply src/recipe.h]: Updated it like that. Old format won't support alternate names or hidden recipes, but still works as before. Only updated one json file as an example.

File: data/json/recipes/food/offal_dishes.json [hunk 6]
```diff
@@ -55,7 +55,10 @@
     "difficulty": 6,
     "charges": 1,
     "time": "30 m",
-    "book_learn": [ [ "mag_glam", 5 ], [ "cookbook_liverforkids", 5, "Buttery Duck Bites" ] ],
+    "book_learn": {
+      "mag_glam": { "skill_level": 5 },
```
Comment (Qrox) [inline data/json/recipes/food/offal_dishes.json]: I'm not sure if using the book item id as the key is a good idea. Usually we put item id together with other data in the same json object, and use a json array to contain all the json objects.

Comment (mqrause) [reply data/json/recipes/food/offal_dishes.json]: I designed it after how it's used in the code. It was a map even before my changes to make sure all entries are unique. This mostly saves the trouble of manually converting everything when reading the JSON. If it's not what you want then I can change it, though I don't really see the problem with it.

**Pull request merged**